File: Kuznetsov_Ilja/1530.py
from typing import List, Tuple, Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def countPairs(self, root: TreeNode, distance: int) -> int:
        ans, _ = Solution.traverse(root, distance)
        return ans
    
    @staticmethod
    def traverse(root: Optional[TreeNode], distance: int) -> Tuple[int, List[int]]:
        default_depths_distr = [0] * (distance + 1)
        if root is None:
            return 0, default_depths_distr
        
        if root.left is None and root.right is None:
            default_depths_distr[1] += 1
            return 0, default_depths_distr
        
        left_pairs, left_depths = Solution.traverse(root.left, distance)
        right_pairs, right_depths  = Solution.traverse(root.right, distance)
        
        good_pairs = left_pairs + right_pairs
        left_partial_sum = 0
        
        for left, right in zip(left_depths, reversed(right_depths)):
            left_partial_sum += left
            good_pairs += right * left_partial_sum
        
        result_distr = [0] + list(map(sum, zip(left_depths, right_depths)))
        result_distr.pop()
        
        return good_pairs, result_distr

# tests
# from math import log2, ceil

# def list_to_root(lst, idx=0):
#     if not idx:
#         lst = lst + [None] * (2 * 2 ** ceil(log2(len(lst))) - len(lst))
    
#     if lst[idx] is None:
#         return None
#     return TreeNode(idx, list_to_root(lst, 2 * idx + 1), list_to_root(lst, 2 * idx + 2))

File: Leetcode/1530.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func DFS(node *TreeNode, distance int, result int ) ([]int, int){
    if (node.Left==nil && node.Right==nil){ // если узел - лист, создаём массив, в котором будет храниться на i позиции кол-во листов с высотой i
        height_count:=make([]int,distance+1)
        height_count[1]+=1
        return height_count, result
    }
    left:=make([]int,distance+1) // массив левой ветки
    right:=make([]int,distance+1) // массив правой ветки
    if (node.Left!=nil){
        left, result=DFS(node.Left, distance, result)
    }
    if (node.Right!=nil){
        right, result=DFS(node.Right, distance, result)
    }
    // проверим, есть ли пары высоты r+l
    for l:=1;l<len(left);l++{
        for r:=distance-1;r>=0;r--{
            if r+l<=distance{
                result+=(left[l]*right[r])
            }
        }
    }
    // для узлов, у которых есть потомки, перепишем массив с высотами
    height_count:=make([]int,distance)
    for i:=len(height_count)-2;i>=1;i--{
        height_count[i+1]=left[i]+right[i]
    }
    return height_count, result
}

func countPairs(root *TreeNode, distance int) int {
    _,result:=DFS(root, distance, 0)
    return result
}